src/librustc_resolve/lib.rs

```diff
@@ -2605,6 +2605,92 @@ impl<'a> Resolver<'a> {
                     mk_struct_span_error(false)
                 };
 
+                err.span_label(ident.span, &format!("this {} is private", binding.res().descr()));
+
+                // Display the chain of re-exports through to the original def for cases where the
+                // `use` is private but the def is public.
+                let mut imported = false;
+                let mut binding = binding;
+                loop {
+                    match binding.kind {
+                        NameBindingKind::Res(res, _is_macro_export) => {
+                            match (res, imported, binding.vis) {
+                                (Res::Def(_, def_id), true, ty::Visibility::Public) => {
+                                    // FIXME: we should verify that this def is actually
+                                    // reachable from the user's crate, as the parent modules
+                                    // of this ADT might be private.
+                                    if def_id.is_local() {
+                                        err.span_label(
+                                            binding.span,
+                                            &format!(
+                                                "consider importing {} `{}` directly",
+                                                res.descr(),
+                                                self.definitions
+                                                    .def_path(def_id.index)
+                                                    .to_string_no_crate(),
```

Comment:
@petrochenkov would you have some advice on how to get a better suited source for a usable path string in `resolve`?

Reply:
There are zero guarantees that a direct path won't have private segments, especially given that module structures with facades are pretty popular.
So trying to guess and use full paths here means inviting false positives.

I'd rather keep the status quo and say only the name.
(In this case the changes in `src/librustc/hir/map/definitions.rs` can be removed as well, those strings are not user-facing otherwise and are used only for debugging.)